srk/src/chc.ml

```diff
@@ -120,120 +176,312 @@ module Fp = struct
     let wg = WeightedGraph.add_vertex (WeightedGraph.empty alg) start_vert in
     let wg = WeightedGraph.add_vertex wg goal_vert in
-    let wg = Relation.Set.fold 
-        (fun rel_sym wg -> WeightedGraph.add_vertex wg rel_sym)
-        (get_relations_used fp)
+    let wg = 
+      List.fold_left 
+        (fun wg rel_sym -> WeightedGraph.add_vertex wg rel_sym)
         wg
+        (get_relations_declared fp)
     in
     let wg = 
       List.fold_left
         (fun wg (rel_atoms, phi, conc) ->
            match rel_atoms with
-           | [] -> WeightedGraph.add_edge 
-                     wg 
-                     start_vert
-                     (emptyarr, BatArray.of_list (params_of_atom conc), phi)
-                     (rel_of_atom conc)
-           | [hd] ->
+           | [] -> 
              WeightedGraph.add_edge 
                wg 
+               start_vert
+               (emptyarr, BatArray.of_list (params_of_atom conc), phi)
+               (rel_of_atom conc)
+           | [hd] -> 
+             WeightedGraph.add_edge 
+               wg
                (rel_of_atom hd)
                (BatArray.of_list (params_of_atom hd), 
                 BatArray.of_list (params_of_atom conc), 
                 phi)
                (rel_of_atom conc)
-           | _ -> failwith "Rule with multiple relations in hypothesis")
+           | _ -> failwith "CHC is non-linear")
         wg
         fp.rules
     in
     let wg =
       List.fold_left
-        (fun wg rel -> 
-           WeightedGraph.add_edge
-             wg
-             rel
-             alg.one
-             goal_vert)
+        (fun wg rel -> WeightedGraph.add_edge wg rel alg.one goal_vert)
         wg
         fp.queries
     in
     wg
 
-  let check srk fp pd = 
-    if is_linear fp then (
-      let wg = to_weighted_graph srk fp pd in
-      let _, _, phi = WeightedGraph.path_weight wg start_vert goal_vert in
-      begin match Smt.is_sat srk phi with
-        | `Unsat -> `No
-        | `Unknown -> `Unknown
-        | `Sat -> `Unknown
-      end)
-    else failwith "No methods for solving non lin fp"
+  let is_super_linear fp =
+    (* Initialize graph: One vertex for each rel symbol.
+     * There is an edge from rel a to rel b if a occurs in the
+     * hypothesis of some rule for which b occurs in the conclusion.
+     * A topological sort on the sccs of this graph will determine the
+     * ordering on which we need to solve the relations.*)
+    let num_rels = (DynArray.length fp.rel_ctx) - 1 in
+    let verts = BatArray.init (num_rels + 1) D.V.create in
+    let graph = D.create () in
+    BatArray.iter (fun v -> D.add_vertex graph v) verts;
+    BatList.iter (fun (hypo_atoms, _, conc_atom) ->
+        List.iter (fun hatom ->
+            D.add_edge 
+              graph 
+              (verts.(rel_of_atom hatom)) 
+              (verts.(rel_of_atom conc_atom)))
+          hypo_atoms)
+      fp.rules;
+    (* We create a new graph in which each scc is collapsed to
```

Comment:
SCCs are [already](http://ocamlgraph.lri.fr/doc/Components.Make.html) ordered topologically, so no need to sort again.  Can check superlin property using scc number.